specification/containerinstance/resource-manager/Microsoft.ContainerInstance/preview/2024-05-01-preview/containerInstance.json

```diff
@@ -0,0 +1,2611 @@
+            "$ref": "#/definitions/ContainerPort"
+          },
+          "x-ms-identifiers": [
+            "port"
+          ]
+        },
+        "environmentVariables": {
+          "description": "The environment variables to set in the container instance.",
+          "type": "array",
+          "items": {
+            "$ref": "#/definitions/EnvironmentVariable"
+          },
+          "x-ms-identifiers": [
+            "name"
+          ]
+        },
+        "instanceView": {
+          "description": "The instance view of the container instance. Only valid in response.",
+          "readOnly": true,
+          "type": "object",
+          "properties": {
+            "restartCount": {
+              "readOnly": true,
+              "type": "integer",
+              "format": "int32",
+              "description": "The number of times that the container instance has been restarted."
+            },
+            "currentState": {
+              "readOnly": true,
+              "description": "Current container instance state.",
+              "$ref": "#/definitions/ContainerState"
+            },
+            "previousState": {
+              "readOnly": true,
+              "description": "Previous container instance state.",
+              "$ref": "#/definitions/ContainerState"
+            },
+            "events": {
+              "readOnly": true,
+              "description": "The events of the container instance.",
+              "type": "array",
+              "items": {
+                "$ref": "#/definitions/Event"
+              },
+              "x-ms-identifiers": [
+                "name",
+                "type"
+              ]
+            }
+          }
+        },
+        "resources": {
+          "description": "The resource requirements of the container instance.",
+          "$ref": "#/definitions/ResourceRequirements"
+        },
+        "volumeMounts": {
+          "description": "The volume mounts available to the container instance.",
+          "type": "array",
+          "items": {
+            "$ref": "#/definitions/VolumeMount"
+          },
+          "x-ms-identifiers": [
+            "name"
+          ]
+        },
+        "livenessProbe": {
+          "description": "The liveness probe.",
+          "$ref": "#/definitions/ContainerProbe"
+        },
+        "readinessProbe": {
+          "description": "The readiness probe.",
+          "$ref": "#/definitions/ContainerProbe"
+        },
+        "securityContext": {
+          "description": "The container security properties.",
+          "$ref": "#/definitions/SecurityContextDefinition"
+        },
+        "configMap": {
+          "description": "The config map.",
+          "$ref": "#/definitions/ConfigMap"
+        }
+      }
+    },
+    "ConfigMap": {
+      "description": "The container config map.",
+      "type": "object",
+      "properties": {
+        "keyValuePairs": {
+          "type": "object",
+          "additionalProperties": {
```

Comment:
>additionalProperties [](http://example.com/codeflow?start=11&length=20)

Use of additionalProperties is not allowed for properties owned by the service. The only time its ok to use it is when the properties are pass thru (user defined) and not subject to any validations. Please make this an array or provide an explanation for why you need this.

Reply:
So, the keyValuePairs object is of type Dictionary<string, string>, since it can be a random collection of user-defined key-value pairs, I have used additionalProperties.

Reply:
The reason this is an ARM anti-pattern is because the properties can change without following ARM versioning. This makes these kinds of APIs difficult and unreliable for customers to use. In addition, since there is no schema, core ARM feature like Azure Policy and Azure Resource Graph can't support them. The exception case is allowed as long as the keys and values are passing through your RP without validation and without being stored. If you need to store, validate, or otherwise use the values, an array is the recommended pattern which is supported by ARM core features. Can you confirm your RP is not using these values directly?

Reply:
RP doesn't use these values.
It doesn't affect control plane in any manner